parlai/core/teachers.py

```diff
@@ -691,22 +694,25 @@ def build_table(self, entry):
 
         :param entry: a tuple in the form described in the class docstring.
         """
-        table = {}
-        if entry[0] is not None:
-            table['text'] = entry[0]
-        if len(entry) > 1:
-            if entry[1] is not None:
-                table['labels'] = entry[1]
-            if len(entry) > 2:
-                if entry[2] is not None:
-                    table['reward'] = entry[2]
-                if len(entry) > 3:
-                    if entry[3] is not None:
-                        table['label_candidates'] = entry[3]
-                    if len(entry) > 4 and entry[4] is not None:
-                        img = self.image_loader.load(entry[4])
-                        if img is not None:
-                            table['image'] = img
+        if isinstance(entry, dict):
+            table = entry
+        else:
+            table = {}
```

Comment:
I'm pretty sure that you don't need the cascading ifs right?

```python
if len(entry) > 0 and entry[0] is not None:
    table['text'] = entry[0]
if len(entry) > 1 and entry[1] is not None:
     table['labels'] = entry[1]
if len(entry) > 2 and entry[2] is not None:
    table['reward'] = entry[2]
# ...
```

really I think you can:

```python
for key, value in zip(['text', 'labels', 'reward', 'label_candidates'], entry):
    if value is not None:
        table[key] = value
# special case on handling images
if len(entry) > 4 and entry[4] is not None:
    table['image'] = self.image_loader.load(entry[4])
```
since zip only iterates through the shortest list:

```
>>> list(zip("abcd", [1, 2, 3]))
[('a', 1), ('b', 2), ('c', 3)]
```

Reply:
brill